test/Microsoft.AspNetCore.Mvc.Core.Test/Controllers/ControllerFeatureProviderTest.cs

```diff
@@ -0,0 +1,546 @@
+        [InlineData(typeof(BasePocoNonControllerController))]
+        [InlineData(typeof(BasePocoNonControllerControllerChild))]
+        [InlineData(typeof(NonController))]
+        [InlineData(typeof(NonControllerChild))]
+        [InlineData(typeof(BaseNonControllerAttributeChildControllerControllerAttributeController))]
+        [InlineData(typeof(PersonModel))] // Verifies that POCO type hierarchies that don't derive from controller return false.
+        public void IsController_ReturnsFalse_IfTypeOrAncestorHasNonControllerAttribute(Type type)
+        {
+            // Arrange
+            var manager = GetApplicationPartManager(type.GetTypeInfo());
+            var feature = new ControllerFeature();
+
+            // Act
+            manager.PopulateFeature(feature);
+
+            // Assert
+            Assert.Empty(feature.Controllers);
+        }
+
+        private static ApplicationPartManager GetApplicationPartManager(params TypeInfo[] types)
+        {
+            var manager = new ApplicationPartManager();
+            manager.ApplicationParts.Add(new TestPart(types));
+            manager.FeatureProviders.Add(new ControllerFeatureProvider());
+
+            return manager;
+        }
+
+        private static IEnumerable<ApplicationPart> GetTestParts(params TypeInfo[] types)
+        {
+            yield return new TestPart(types);
+        }
+
+        private class TestPart : ApplicationPart, IApplicationPartTypeProvider
+        {
+            public TestPart(params TypeInfo[] types)
+            {
+                Types = types;
+            }
+
+            public override string Name => "Test";
+
+            public IEnumerable<TypeInfo> Types { get; }
+        }
+    }
+}
+
+// These controllers are used to test the ControllerFeatureProvider implementation
+// which REQUIRES that they be public top-level classes. To avoid having to stub out the
+// implementation of this class to test it, they are just top level classes. Don't reuse
+// these outside this test - find a better way or use nested classes to keep the tests
+// independent.
+namespace Microsoft.AspNetCore.Mvc.ControllerFeatureProviderControllers
+{
+    public abstract class AbstractController : Controller
+    {
+    }
+
+    public class DerivedAbstractController : AbstractController
+    {
+    }
+
+    public class StoreController : Controller
+    {
+    }
+
+    public class ProductsController : ControllerBase
+    {
+    }
+
+    public class Products : ControllerBase
+    {
+    }
+
+    [Controller]
+    public abstract class Controller
+    {
+    }
+
+    public abstract class NoControllerAttributeBaseController
+    {
+    }
+
+    public class NoSuffixNoControllerAttribute : NoControllerAttributeBaseController
+    {
+    }
+
+    public class OpenGenericController<T> : Controller
+    {
+    }
+
+    public class DerivedGenericController : OpenGenericController<string>
+    {
+    }
+
+    public interface IController
```

Comment:
I see this is just cut & paste, but this confused me to no end at first. If we rename this to `ISomeController` would it still have the same value as a test? (From the looks of it yes.)